### PR TITLE
Enable Flan-T5 inference in `float16` 

### PR DESCRIPTION
**Pull request by HMellor**
# What does this PR do?

- Enable floating point exceptions (probably temporarily) so we can track down the source of over/underflows 
- Fix overflow coming from attention masking
- Fix overflow coming from tanh approximation of GeLU
- Fix overflow coming from pre-norm residual structure
- Fix overflow coming from FeedForward down projection

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

File: optimum/graphcore/models/t5/modeling_t5.py [hunk 1]
```diff
@@ -310,15 +313,24 @@ def deparallelize(self):
...
+            block.layer[0].dropout = block.layer[0].dropout.module
+            with torch.no_grad():
+                block.layer[1].DenseReluDense.wo.weight *= block.layer[1].dropout.scale
+            block.layer[1].dropout = block.layer[1].dropout.module
+            if self.config.dense_act_fn == "gelu_new":
+                block.layer[1].DenseReluDense.act = NewGELUActivation
         for block in self.decoder.block:
             block.__class__ = T5Block
+            if self.config.dense_act_fn == "gelu_new":
+                block.layer[1].DenseReluDense.act = NewGELUActivation
```
Comment (kundaMwiza) [inline optimum/graphcore/models/t5/modeling_t5.py:333]: This should be `block.layer[2]`

File: optimum/graphcore/models/t5/modeling_t5.py [hunk 2]
```diff
@@ -310,15 +313,24 @@ def deparallelize(self):
...
+        self.encoder.embed_tokens = self.encoder.embed_tokens.module
+
         for block in self.encoder.block:
             block.__class__ = T5Block
+            block.layer[0].dropout = block.layer[0].dropout.module
+            with torch.no_grad():
+                block.layer[1].DenseReluDense.wo.weight *= block.layer[1].dropout.scale
+            block.layer[1].dropout = block.layer[1].dropout.module
+            if self.config.dense_act_fn == "gelu_new":
+                block.layer[1].DenseReluDense.act = NewGELUActivation
```
Comment (kundaMwiza) [inline optimum/graphcore/models/t5/modeling_t5.py:329]: The activation functions `newGELUActivation` should be instantiated

File: optimum/graphcore/ipu_configuration.py [hunk 3]
```diff
@@ -290,6 +290,9 @@ def _to_options(self, for_inference: bool = False, compile_only: bool = False) -
 
         opts._Popart.set("saveInitializersToFile", NamedTemporaryFile().name)
 
+        # Enable floating point exceptions
+        opts.Precision.enableFloatingPointExceptions(True)
+
```
Comment (HMellor) [inline optimum/graphcore/ipu_configuration.py:295]:
```suggestion
```

**Review by jimypbr — APPROVED** [review]
